Added "tpl" template function

eicnix opened the pull request:
Added implementation for a tpl function that evaluates a string as a template. 

Implementation for #1978

eicnix committed:
Added implementation for a tpl function that evaluates a string as a template.

Implementation for #1978

technosophos commented on pkg/engine/engine_test.go:
```diff
@@ -445,4 +445,33 @@ func TestAlterFuncMap(t *testing.T) {
+	tplChart := &chart.Chart{
+		Metadata: &chart.Metadata{Name: "TplFunction"},
+		Templates: []*chart.Template{
+			{Name: "templates/base", Data: []byte(`Evaluate tpl {{tpl "Value: {{ .Values.value}}" .}}`)},
```
This looks great!

Can you also test a template that uses template functions (like `quote`) and defined templates (`{{ define "foo" }}`)? Something like `tpl "Value: {{ .Values.value | quote }}"` and `tpl "{{ include 'foo' }}"`.

It looks like those should both work. But I think it's worth having a safety net in the tests. We need to be sure that now and in the future those two items are available within a nested template.

eicnix replied:
While testing the "include" function I ran into the following error since the other template is only available for the outer render context. Is there an global render context I could use to pass all registered templates to the rendering of the "tpl"-function? 

```
tplChartWithInclude := &chart.Chart{
		Metadata: &chart.Metadata{Name: "TplFunction"},
		Templates: []*chart.Template{
			{Name: "templates/base", Data: []byte(`{{ tpl "{{include `+"`"+`TplFunction/templates/_partial`+ "`" + ` . | indent 2}}" .}}`)},
			{Name: "templates/_partial", Data: []byte(`{{.Release.Name}} - he`)},
		},
		Values:       &chart.Config{Raw: ``},
		Dependencies: []*chart.Chart{},
```

results in: 

`Expected "Evaluate tpl bar", got "  template: no template \"TplFunction/templates/_partial\" associated with template \"gotpl\"" (map[TplFunction/templates/_partial:TestRelease - he TplFunction/templates/base:  template: no template "TplFunction/templates/_partial" associated with template "gotpl"])`

technosophos replied:
I'll look into it. I was wondering about the scoping for that.

databus23 commented on pkg/engine/engine.go:
```diff
@@ -158,6 +161,23 @@ func (e *Engine) alterFuncMap(t *template.Template) template.FuncMap {
+		templates["template"] = r
+
+		result, err := e.render(templates)
+		if err != nil {
```
Instead of returning the error as the result string I would favour making the function return `(string, error)` and failing the template execution altogether if the provided string does not render.

Otherwise errors are very hard to catch and will most certainly endup being deployed to k8s causing failures further down the road.

See the recently merged change for `include` https://github.com/kubernetes/helm/pull/2358 for some more details.

eicnix replied:
Thanks for the input. I changed the function to return an error on failure instead of writing it into the template.

eicnix committed:
- Changed error behaviour to returning an error instead of writing the error in the template
- Added tests for using a function inside a template that is evaluated using the "tpl" function

technosophos commented on pkg/engine/engine.go:
```diff
@@ -158,6 +161,23 @@ func (e *Engine) alterFuncMap(t *template.Template) template.FuncMap {
+		templates := map[string]renderable{}
+		templates["template"] = r
+
+		result, err := e.render(templates)
```
I wonder if doing a `t.Parse()` here instead might cause it to inherit scope. I'm not sure how best to work around the loss of context...

eicnix replied:
When a chart is rendered I add the Chart's templates as the current templates in the engine. If something else is rendered it will add the current templates to the render context.

@technosophos What do you think about it?

eicnix committed:
Added storage of parsed templates in engine

eicnix committed:
Added unit test for include in tpl function

eicnix committed:
removed debug file

technosophos commented:
Is this ready for another review, @eicnix ?

eicnix committed:
Merge branch 'master' into master

eicnix commented:
@technosophos Yes, I am saving the current templates in the engine to reuse them in a `tpl` execution. 
A test with uses an `include` inside a `tpl` is working.

technosophos commented:
Okay. It's on my docket for testing tomorrow for inclusion in 2.5.0.

technosophos commented:
Okay, I have done a lot of testing of this, and it seems to be in working order. Do you feel okay with my documenting this as an experimental feature for now? I am still a little worried that some strange edge case will crop up in production. But I am confident enough in this PR that I think we should ship it for 2.5.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/helm/2350)
<!-- Reviewable:end -->

technosophos commented on pkg/engine/engine.go:
```diff
@@ -159,6 +164,23 @@ func (e *Engine) alterFuncMap(t *template.Template) template.FuncMap {
+		}
+
+		templates := map[string]renderable{}
+		templates["aaa_template"] = r
```
I would not mind changing this to a randomly generated string just to discourage others from trying to access the template directly. But that's definitely not necessary for now.

technosophos commented:
Okay! I'm merging this. Thanks again for the work on this @eicnix

scjody commented:
@eicnix Would it be possible for you to add some docs on what the `tpl` function does? As a Helm user I had trouble figuring this out and had to get help on Slack.

eicnix commented:
@scjody I added some documentation at https://github.com/kubernetes/helm/pull/4160 . Would this documentation have been sufficient for you?

scjody commented:
Yes, that looks great, thanks!
